### PR TITLE
Migrate `pallet-treasury` to the new pallet attribute macro

### PR DESCRIPTION
**Pull request by koushiro**
Part of #7882

Migrate the `pallet-treasury` to the new pallet attribute macro.

## :warning:  Breaking Change :warning:

From https://crates.parity.io/frame_support/attr.pallet.html#checking-upgrade-guidelines

> storages now use PalletInfo for module_prefix instead of the one given to decl_storage: Thus any use of this pallet in construct_runtime! should be careful to update name in order not to break storage or to upgrade storage (moreover for instantiable pallet). If pallet is published, make sure to warn about this breaking change.

So users of the pallet must be careful about the name they used in `construct_runtime!`. Hence the `runtime-migration` label, which might not be needed depending on the configuration of the pallet.

`polkadot` and `kusama` use `Treasury` as pallet name, thus no need for migration

---

polkadot companion: paritytech/polkadot#3359

### CODE REVIEW

**Comment by koushiro** [comment]
@thiolliere @KiChjang PTAL

File: frame/treasury/src/lib.rs [hunk 1]
```diff
@@ -175,92 +128,182 @@ pub struct Proposal<AccountId, Balance> {
...
+	impl Default for GenesisConfig {
+		fn default() -> Self {
+			Self
+		}
+	}
+
+	#[cfg(feature = "std")]
+	impl GenesisConfig {
+		/// Direct implementation of `GenesisBuild::assimilate_storage`.
+		pub fn assimilate_storage<T: Config<I>, I: 'static>(
```
Comment (thiolliere) [inline frame/treasury/src/lib.rs:230]: I think we should deprecate this direct implementation, I know it was generated by decl_storage, but people shouldn't really need it if they use construct_runtime and the runtime GensisConfig build.
```suggestion
		#[deprecated(note = "use `<GensisConfig<T, I> as GenesisBuild<T, I>>::assimilate_storage` instead")]
		pub fn assimilate_storage<T: Config<I>, I: 'static>(
```

File: frame/treasury/src/lib.rs [hunk 2]
```diff
@@ -175,92 +128,182 @@ pub struct Proposal<AccountId, Balance> {
...
+
+		/// Period between successive spends.
+		#[pallet::constant]
+		type SpendPeriod: Get<Self::BlockNumber>;
+
+		/// Percentage of spare funds (if any) that are burnt per spend period.
+		#[pallet::constant]
+		type Burn: Get<Permill>;
+
+		/// The treasury's module id, used for deriving its sovereign account ID.
```
Comment (thiolliere) [inline frame/treasury/src/lib.rs:175]:
```suggestion
		/// The treasury's pallet id, used for deriving its sovereign account ID.
```

**Review by thiolliere — APPROVED** [review]
looks good, polkadot and kusama does use the name `Treasury` for the name of the pallet, thus no migration needed.
westend and rococo don't have treasury pallet

**Review by shawntabrizi — APPROVED** [review]

**Comment by shawntabrizi** [comment]
bot merge

**Comment by shawntabrizi** [comment]
bot merge

**Pull request merged**